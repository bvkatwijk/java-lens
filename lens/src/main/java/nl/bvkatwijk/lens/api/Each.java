package nl.bvkatwijk.lens.api;

import io.vavr.Value;

import java.util.function.UnaryOperator;

public class Each {
    public static <V extends Value<T>, T> UnaryOperator<V> map(UnaryOperator<T> f) {
        return it -> (V) it.map(f);
    }
}
