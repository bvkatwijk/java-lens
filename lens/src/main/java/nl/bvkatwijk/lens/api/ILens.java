package nl.bvkatwijk.lens.api;

import nl.bvkatwijk.lens.ops.GetOps;
import nl.bvkatwijk.lens.ops.ModifyOps;
import nl.bvkatwijk.lens.ops.WithOps;

import java.util.function.Function;

public interface ILens<S, T> extends ModifyOps<S, T>, GetOps<S, T>, WithOps<S, T> {
    default <U> ILens<S, U> andThen(ILens<T, U> lens) {
        return new Lens<>(
            get().andThen(lens::get),
            (s, u) -> with(s, lens.with(get(s), u)));
    }

    /**
     * Compose target lens with this one, returning a new lens that first applies the argument lens
     * before applying this one.
     * <br />
     * See {@link java.util.function.Function#compose(Function)}
     *
     * @param lens target {@link ILens}
     * @return {@link ILens} instance applying target lens before applying this one.
     * @param <R> Source type
     */
    default <R> ILens<R, T> compose(ILens<R, S> lens) {
        return lens.andThen(this);
    }
}
