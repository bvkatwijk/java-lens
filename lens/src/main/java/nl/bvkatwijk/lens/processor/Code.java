package nl.bvkatwijk.lens.processor;

import io.vavr.collection.List;
import io.vavr.collection.Vector;
import nl.bvkatwijk.lens.Const;

public class Code {
    static String params(String... args) {
        return Vector.of(args)
            .mkString(", ");
    }

    /**
     * @param typeName type name
     * @param method method name
     * @return method reference
     */
    static String reference(String typeName, String method) {
        return typeName + "::" + method;
    }

    static String importStatement(String qualifiedTypeName) {
        return "import " + qualifiedTypeName + ";";
    }

    static String importStatements(String... args) {
        return Vector.of(args)
            .map(Code::importStatement)
            .mkString("\n");
    }

    static String iLens(String from, String to) {
        return Const.ILENS + "<" + params(from, to) + ">";
    }

    static String iLens(String name) {
        return iLens(Const.PARAM_SOURCE_TYPE, name);
    }

    static String unqualify(String qualifiedType) {
        return qualifiedType.substring(qualifiedType.lastIndexOf(".") + 1);
    }

    static String indent(String string) {
        return Const.INDENT + string;
    }

    static List<String> indent(List<String> strings) {
        return strings.map(Code::indent);
    }
}
