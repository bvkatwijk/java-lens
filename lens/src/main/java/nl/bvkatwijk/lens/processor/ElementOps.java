package nl.bvkatwijk.lens.processor;

import javax.lang.model.element.Element;
import javax.lang.model.element.RecordComponentElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;

/**
 * Utiilty functions on {@link javax.lang.model.element.Element} and subclasses
 */
public final class ElementOps {
    static String fieldName(RecordComponentElement it) {
        return it.getSimpleName().toString();
    }

    static String typeName(Element it) {
        TypeMirror type = it.asType();
        return switch (type.getKind()) {
            case BOOLEAN -> Boolean.class.getName();
            case BYTE -> Byte.class.getName();
            case SHORT -> Short.class.getName();
            case INT -> Integer.class.getName();
            case LONG -> Long.class.getName();
            case CHAR -> Character.class.getName();
            case FLOAT -> Float.class.getName();
            case DOUBLE -> Double.class.getName();
            case VOID -> Void.class.getName(); // Does it make sense to support Void type?
            case DECLARED -> type.toString();
            case OTHER, NONE, MODULE, INTERSECTION, UNION, EXECUTABLE, PACKAGE, WILDCARD, TYPEVAR, ERROR, ARRAY, NULL ->
                throw new IllegalArgumentException("Type " + it + " (" + it.getKind() + " " + type.getKind() + ") not yet supported.");
        };
    }

    static String unqualifiedTypeName(RecordComponentElement element) {
        return Code.unqualify(typeName(element));
    }

    static String packageElement(Element element) {
        return switch (element.getKind()) {
            case PACKAGE -> element.toString();
            case RECORD_COMPONENT -> {
                if (element.asType() instanceof DeclaredType declaredType) {
                    yield packageElement(declaredType.asElement());
                }
                yield "can this happen? primitives maybe?";
            }
            default -> packageElement(element.getEnclosingElement());
        };
    }
}
