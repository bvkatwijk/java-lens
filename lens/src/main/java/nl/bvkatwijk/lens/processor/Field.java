package nl.bvkatwijk.lens.processor;

import javax.lang.model.element.RecordComponentElement;

public record Field(String typeName, String fieldName, RecordComponentElement unused) {
}
