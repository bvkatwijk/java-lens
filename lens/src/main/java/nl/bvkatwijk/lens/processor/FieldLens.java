package nl.bvkatwijk.lens.processor;

import io.vavr.collection.List;
import nl.bvkatwijk.lens.Const;

import javax.lang.model.element.RecordComponentElement;

record FieldLens(String qualifiedType, LensKind lensKind, RecordComponentElement field) {
    public List<String> lensMethod() {
        return Code.indent(List.of(
            "",
            "public " + returnType() + " " + Code.fieldName(field) + "() {",
            Code.indent(returnStatement()),
            "}"
        ));
    }

    public String returnType() {
        return switch (lensKind) {
            case LENSED -> typeLens(field) + "<" + Const.PARAM_SOURCE_TYPE + ">";
            case PRIMITIVE, OTHER -> LensCode.iLens(qualifiedType);
        };
    }

    public String returnStatement() {
        var chainInner = "inner.andThen(" + LensCode.lensName(field) + ")";
        return Code.ret(switch (lensKind) {
            case LENSED -> "new " + typeLens(field) + "<>(" + chainInner + ")";
            case PRIMITIVE, OTHER -> chainInner;
        });
    }

    static String typeLens(RecordComponentElement element) {
        return LensProcessor.packageElement(element) + "." + fieldTypeUnqualified(element) + Const.LENS;
    }

    static String fieldTypeUnqualified(RecordComponentElement it) {
        return Code.unqualify(it.asType().toString());
    }
}
