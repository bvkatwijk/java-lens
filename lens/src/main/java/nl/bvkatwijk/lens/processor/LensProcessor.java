package nl.bvkatwijk.lens.processor;

import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import lombok.SneakyThrows;
import nl.bvkatwijk.lens.Const;
import nl.bvkatwijk.lens.Lenses;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.RecordComponentElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import java.io.IOException;
import java.util.Set;
import java.util.regex.Pattern;

@SupportedSourceVersion(SourceVersion.RELEASE_19)
@SupportedAnnotationTypes("nl.bvkatwijk.lens.Lenses")
public class LensProcessor extends AbstractProcessor {
    protected static final String LENS = "Lens";

    @Override
    @SneakyThrows
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        if (annotations.isEmpty() && roundEnv.processingOver()) {
            return true;
        }
        lensElements(roundEnv)
            .filter(it -> ElementKind.RECORD.equals(it.getKind()))
            .forEach(this::writeSourceFile);
        return true;
    }

    @SneakyThrows
    private void writeSourceFile(Element el) {
        var pack = "nl.bvkatwijk.lens.gen";
        var name = el.getSimpleName().toString();
        var fields = List.ofAll(el.getEnclosedElements()
            .stream()
            .filter(RecordComponentElement.class::isInstance)
            .map(RecordComponentElement.class::cast)
            .toList());
        var isoConstants = fields
            .map(it -> isoConstant(name, it.getSimpleName().toString(), typeName(it)))
            .map(this::indent)
            .toList();
        writeSourceFile(pack, name, String.join(
            "\n",
            List.of("package " + pack + ";", "")
                .append(importLens())
                .appendAll(imports(List.of(el).appendAll(fields)))
                .append("")
                .append("public class " + name + Const.LENS + " {")
                .appendAll(isoConstants)
                .append("}")
                .toJavaList()));
    }

    private String importLens() {
        return "import nl.bvkatwijk.lens.kind.Lens;";
    }

    private Iterable<String> imports(List<? extends Element> fields) {
        return fields
            .map(LensProcessor::typeName)
            .map(it -> it.indexOf('<') > 0 ? it.substring(0, it.indexOf('<')) : it)
            .map(LensProcessor::importElement);
    }

    private static String importElement(String qualifiedTypeName) {
        return "import " + qualifiedTypeName + ";";
    }

    private static String typeName(Element it) {
        TypeMirror type = it.asType();
        return switch (type.getKind()) {
            case BOOLEAN -> "java.lang.Boolean";
            case BYTE -> "java.lang.Byte";
            case SHORT -> "java.lang.Short";
            case INT -> "java.lang.Integer";
            case LONG -> "java.lang.Long";
            case CHAR -> "java.lang.Character";
            case FLOAT -> "java.lang.Float";
            case DOUBLE -> "java.lang.Double";
            case VOID -> "java.lang.Void";
            case DECLARED, OTHER -> type.toString();
            case NONE, MODULE, INTERSECTION, UNION, EXECUTABLE, PACKAGE, WILDCARD, TYPEVAR, ERROR, ARRAY, NULL -> throw new IllegalArgumentException("Type " + it + " (" + it.getKind() + " " + type.getKind() + ") not yet supported.");
        };
    }

    public String isoConstant(String record, String field, String fieldType) {
        return "public static final " + Const.LENS + "<" + record + ", " + fieldType + "> " + isoName(field) + " = new " + Const.LENS + "<>(" + record + "::" + witherName(
            field) + ", " + record + "::" + field + ");";
    }

    private static String isoName(String field) {
        return field.toUpperCase();
    }

    public String witherName(String fieldName) {
        return "with" + capitalize(fieldName);
    }

    public String indent(String string) {
        return Const.INDENT + string;
    }

    public String capitalize(String string) {
        return Pattern.compile("^.")
            .matcher(string)
            .replaceFirst(m -> m.group().toUpperCase());
    }

    public List<String> indent(List<String> strings) {
        return strings.map(this::indent);
    }

    private void writeSourceFile(String pack, String name, String content) throws IOException {
        processingEnv.getFiler()
            .createSourceFile(pack + "." + name + LENS)
            .openWriter()
            .append(content)
            .close();
    }

    private static HashSet<? extends Element> lensElements(RoundEnvironment roundEnv) {
        return HashSet.ofAll(roundEnv.getElementsAnnotatedWith(Lenses.class));
    }
}
