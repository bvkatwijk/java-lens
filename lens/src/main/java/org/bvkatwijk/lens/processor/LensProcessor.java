package org.bvkatwijk.lens.processor;

import io.vavr.collection.HashSet;
import lombok.SneakyThrows;
import org.bvkatwijk.lens.Lens;
import org.bvkatwijk.lens.ast.Type;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import java.io.IOException;
import java.util.Set;

@SupportedSourceVersion(SourceVersion.RELEASE_19)
@SupportedAnnotationTypes("org.bvkatwijk.lens.Lens")
public class LensProcessor extends AbstractProcessor {
    @Override
    @SneakyThrows
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        if (annotations.isEmpty() && roundEnv.processingOver()) {
            return true;
        }
        lensElements(roundEnv)
            .filter(it -> ElementKind.RECORD.equals(it.getKind()))
            .forEach(it -> writeSourceFile(new Type("org.bvkatwijk.lens.gen", it.getSimpleName() + "Lens")));
        return true;
    }

    @SneakyThrows
    private void writeSourceFile(Type it) {
        writeSourceFile(it, String.join(
            "\n",
            "package " + it.pack() + ";",
            "",
            "import org.bvkatwijk.lens.Address;",
            "import org.bvkatwijk.lens.Focus;",
            "import org.bvkatwijk.lens.IFocus;",
            "",
            "import java.util.function.Function;",
            "",
            "public class " + it.name() + " {",
            "    public <TARGET> IFocus<Address, TARGET> focus(Function<Address, TARGET> f, Function<TARGET, Address> with) {",
            "        return new Focus<>(this, with, f);",
            "    }",
            "}"
        ));
    }

    private void writeSourceFile(Type it, String content) throws IOException {
        processingEnv.getFiler()
            .createSourceFile(it.qualified())
            .openWriter()
            .append(content)
            .close();
    }

    private static HashSet<? extends Element> lensElements(RoundEnvironment roundEnv) {
        return HashSet.ofAll(roundEnv.getElementsAnnotatedWith(Lens.class));
    }
}
