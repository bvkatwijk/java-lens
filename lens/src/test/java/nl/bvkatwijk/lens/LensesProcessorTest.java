package nl.bvkatwijk.lens;

import org.junit.jupiter.api.Test;

public class LensesProcessorTest {
    @Test
    void verifyProcessed() {
    }
}
