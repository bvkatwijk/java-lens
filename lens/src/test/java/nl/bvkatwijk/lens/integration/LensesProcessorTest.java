package nl.bvkatwijk.lens.integration;

import com.google.testing.compile.Compilation;
import lombok.SneakyThrows;
import nl.bvkatwijk.lens.processor.LensProcessor;
import nl.bvkatwijk.lens.test.TestUtil;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import javax.tools.JavaFileObject;

public class LensesProcessorTest {
    @Test
    @SneakyThrows
    void processedRecordComponentElement() {
        var lensProcessor = new LensProcessor();
        JavaFileObject load = TestUtil.load("example/Person.java");
        Compilation compile = TestUtil.compile(lensProcessor, load);
        Approvals.verify(compile.generatedSourceFiles()
            .getFirst()
            .getCharContent(true)
            .toString());
    }
}
