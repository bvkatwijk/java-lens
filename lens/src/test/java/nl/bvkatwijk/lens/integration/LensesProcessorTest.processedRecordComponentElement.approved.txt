package nl.bvkatwijk.lens.example;

import nl.bvkatwijk.lens.example.Person;
import nl.bvkatwijk.lens.api.ILens;
import nl.bvkatwijk.lens.api.Lens;

public record PersonLens<SOURCE>(ILens<SOURCE, Person> inner) implements ILens<SOURCE, Person> {
	public static final PersonLens<Person> Âµ = new PersonLens<>(Lens.identity());
	public static final ILens<Person, java.lang.String> NAME = new Lens<>(Person::name, Person::withName);
	public static final ILens<Person, nl.bvkatwijk.lens.example.Address> ADDRESS = new Lens<>(Person::address, Person::withAddress);
	public static final ILens<Person, nl.bvkatwijk.lens.example.Address> WORK = new Lens<>(Person::work, Person::withWork);
	public static final ILens<Person, io.vavr.collection.List<nl.bvkatwijk.lens.example.Person>> FRIENDS = new Lens<>(Person::friends, Person::withFriends);
	public static final ILens<Person, java.lang.Boolean> COOL = new Lens<>(Person::cool, Person::withCool);
	
	public ILens<SOURCE, java.lang.String> name() {
		return inner.andThen(NAME);
	}
	
	public nl.bvkatwijk.lens.example.AddressLens<SOURCE> address() {
		return new nl.bvkatwijk.lens.example.AddressLens<>(inner.andThen(ADDRESS));
	}
	
	public nl.bvkatwijk.lens.example.AddressLens<SOURCE> work() {
		return new nl.bvkatwijk.lens.example.AddressLens<>(inner.andThen(WORK));
	}
	
	public ILens<SOURCE, io.vavr.collection.List<nl.bvkatwijk.lens.example.Person>> friends() {
		return inner.andThen(FRIENDS);
	}
	
	public ILens<SOURCE, java.lang.Boolean> cool() {
		return inner.andThen(COOL);
	}
	
	public java.util.function.BiFunction<SOURCE, Person, SOURCE> with() {
		return inner.with();
	}
	
	public java.util.function.Function<SOURCE, Person> get() {
		return inner.get();
	}
}